node/pkg/algorand/watcher.go

```diff
@@ -2,29 +2,254 @@ package algorand
+	})
+
+	logger := supervisor.Logger(ctx)
+	errC := make(chan error)
+
+	logger.Info("Algorand watcher connecting", zap.String("url", e.indexerRPC))
+
+	go func() {
+		timer := time.NewTicker(time.Second * 1)
+		defer timer.Stop()
+
+		indexerClient, err := indexer.MakeClient(e.indexerRPC, e.indexerToken)
+		if err != nil {
+			logger.Info("indexer make client", zap.Error(err))
+			p2p.DefaultRegistry.AddErrorCount(vaa.ChainIDAlgorand, 1)
+			errC <- err
+			return
+		}
+
+		// Parameters
+		var notePrefix = "publishMessage"
 
-	select {}
+		if e.next_round == 0 {
+			// What is the latest round...
+			result, err := indexerClient.SearchForTransactions().Limit(0).Do(context.Background())
+			if err != nil {
+				logger.Info("SearchForTransaction", zap.Error(err))
+				p2p.DefaultRegistry.AddErrorCount(vaa.ChainIDAlgorand, 1)
+				errC <- err
+				return
+			}
+			e.next_round = result.CurrentRound + 1
+			logger.Info(fmt.Sprintf("Algorand next_round set to %d", e.next_round))
+		}
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case r := <-e.obsvReqC:
+				if vaa.ChainID(r.ChainId) != vaa.ChainIDAlgorand {
+					panic("invalid chain ID")
+				}
+
+				logger.Info("Received obsv request",
+					zap.String("tx_hash", hex.EncodeToString(r.TxHash)),
+					zap.String("base32_tx_hash", base32.StdEncoding.WithPadding(base32.NoPadding).EncodeToString(r.TxHash)))
+
+				result, err := indexerClient.SearchForTransactions().TXID(base32.StdEncoding.WithPadding(base32.NoPadding).EncodeToString(r.TxHash)).Do(context.Background())
+				if err != nil {
+					logger.Info("SearchForTransactions", zap.Error(err))
+					p2p.DefaultRegistry.AddErrorCount(vaa.ChainIDAlgorand, 1)
+					errC <- err
+					return
+				}
+				for i := 0; i < len(result.Transactions); i++ {
+					var t = result.Transactions[i]
+					lookAtTxn(e, t, logger)
+				}
+
+			case <-timer.C:
+				var nextToken = ""
+				for {
+					result, err := indexerClient.SearchForTransactions().NotePrefix([]byte(notePrefix)).MinRound(e.next_round).NextToken(nextToken).Do(context.Background())
+					if err != nil {
+						logger.Info("SearchForTransaction", zap.Error(err))
+						p2p.DefaultRegistry.AddErrorCount(vaa.ChainIDAlgorand, 1)
+						errC <- err
+						return
+					}
+
+					for i := 0; i < len(result.Transactions); i++ {
+						var t = result.Transactions[i]
+						lookAtTxn(e, t, logger)
+					}
+
+					if result.NextToken != "" {
+						nextToken = result.NextToken
+					} else {
+						e.next_round = result.CurrentRound + 1
+						break
+					}
+				}
+				readiness.SetReady(common.ReadinessAlgorandSyncing)
+				currentAlgorandHeight.Set(float64(e.next_round - 1))
+				p2p.DefaultRegistry.SetNetworkStats(vaa.ChainIDAlgorand, &gossipv1.Heartbeat_Network{
+					Height:          int64(e.next_round - 1),
+					ContractAddress: fmt.Sprintf("%d", e.appid),
+				})
+			}
+		}
+	}()
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-errC:
+		return err
+	}
```

Comment:
The other watchers fetch the guardian set periodically (every 15 seconds). I don't see that here? Is that intentional?

Reply:
I guess I don't know? If you heard from an authoritative source, then I would go with that?

Reply:
https://github.com/certusone/wormhole/blob/76f8acf9ec07df51898e6ff3240c2b12428b293f/node/pkg/terra/watcher.go#L305